core/src/main/java/io/undertow/server/MultiValueHashListStorage.java

```diff
@@ -0,0 +1,170 @@
+ *  limitations under the License.
+ */
+package io.undertow.server;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Set;
+
+/**
+ * This class allows to store multiple values for single key value. Key and value order are honored.
+ * Duplicate entries of key-value pair is not allowed, it is vetted via {@link java.lang.Object#equals()} method.
+ * @author baranowb
+ *
+ * @param <K> - key type
+ * @param <V> - value type
+ */
+public class MultiValueHashListStorage<K, V> {
+    // NOTE: requests are processed by single thread so no need to sync?
+    private LinkedHashMap<K, ArrayList<V>> storage = new LinkedHashMap<>();
+
+    public void put(K key, V value) {
+        ArrayList<V> lst = storage.get(key);
+        if (lst == null) {
+            lst = new ArrayList<>();
+            lst.add(value);
+            storage.put(key, lst);
+        } else {
+            if (lst.contains(value)) {
+                lst.remove(value);
+            }
+            lst.add(value);
+        }
+    }
+
+    public Iterator<V> valuesIterator() {
+
+        return new DefaultIterator<V>(this.storage);
+    }
+
+    public int size() {
+        int size = 0;
+        for (Entry<K, ArrayList<V>> kvp : storage.entrySet()) {
+            final List<V> lst = kvp.getValue();
+            if (lst != null) {
+                size += lst.size();
+            }
+        }
+        return size;
+    }
+
+    public boolean isEmpty() {
+        return storage.isEmpty();
+    }
+
+    public boolean containsKey(Object key) {
+        return storage.containsKey(key);
+    }
+
+    public boolean containsValue(Object value) {
+        for (Entry<K, ArrayList<V>> kvp : storage.entrySet()) {
+            final List<V> lst = kvp.getValue();
+            if (lst != null && lst.contains(value)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public ArrayList<V> get(K key) {
+        final ArrayList<V> values = storage.get(key);
+        return values != null ? values : new ArrayList<V>();
+    }
+
+    public Collection<V> removeAll(Object key) {
+        final ArrayList<V> values = storage.remove(key);
+        return values != null ? values : new ArrayList<V>();
+    }
+
+    @SuppressWarnings("unchecked")
+    public V remove(Object key, Object value) {
+        final ArrayList<V> values = storage.get(key);
+        if (values.contains(value)) {
+            values.remove(value);
+            if (values.size() == 0) {
+                storage.remove(key);
+            }
+            return (V) value;
+        } else {
+            return null;
+        }
+    }
+
+    public Set<K> keySet() {
+        return storage.keySet();
+    }
+
+    public void clear() {
+        storage.clear();
+    }
+
+    private class DefaultIterator<V> implements Iterator<V> {
```

Comment:
DefaultIterator should have ideally both "static final" classifiers.  With "static" classifier  DefaultIterator class could  be shared among all MultiValueHashListStorage object instances. Without "static" classifier new DefaultIterator class is created per MultiValueHashListStorage instances. This change will result in improvements to shared memory AKA "method area" .